oci/spec_opts.go

```diff
@@ -393,12 +396,77 @@ func WithImageConfigArgs(image Image, args []string) SpecOpts {
 			// even if there is no specified user in the image config
 			return WithAdditionalGIDs("root")(ctx, client, c, s)
 		} else if s.Windows != nil {
+			// imageExtended is a subset of the oci Image struct that changes
+			// the Config type to be imageConfigExtended in order to add the
+			// ability to deserialize `ArgsEscaped` which is not an OCI field,
+			// but is supported by Docker built images.
+			type imageExtended struct {
+				Config struct {
+					ArgsEscaped bool `json:"ArgsEscaped,omitempty"`
+				}
+			}
+			// Deserialize the extended image format for Windows.
+			var ociImageExtended imageExtended
+			if err := json.Unmarshal(imageConfigBytes, &ociImageExtended); err != nil {
+				return err
+			}
+			argsEscaped := ociImageExtended.Config.ArgsEscaped
+
 			s.Process.Env = replaceOrAppendEnvValues(config.Env, s.Process.Env)
-			cmd := config.Cmd
-			if len(args) > 0 {
-				cmd = args
+
+			if argsEscaped {
+				// For certain Windows images created via Docker they may
+				// contain `ArgsEscaped==true` if the ENTRYPOINT or CMD is in
+				// the `shell` form and contains spaces, path characters, etc.
+				// An example is the following:
+				//
+				// ```
+				// FROM mcr.microsoft.com/windows/servercore:ltsc2019
+				//
+				// SHELL ["powershell", "-Command"]
+				// ENTRYPOINT "C:\Path To\MyExe.exe" -arg1 "-arg2 value2"
+				// ```
+				//
+				// Will be encoded by Docker as:
+				// ```
+				// "Config": {
+				//   ...
+				//   "ArgsEscaped": true,
+				//   ...
+				//   "Entrypoint": [
+				//     "powershell.exe -Command \"C:\\My Data\\MyExe.exe\" -arg1 \"-arg2 value2\""
+				//   ],
+				//   ...
+				// }
+				// ```
+				//
+				// You can see that `Entrypoint` becomes a single element array
+				// with the entire argument list already escaped. To avoid a
+				// double escape problem Windows supports passing via OCI as a
+				// complete `CommandLine`.
+				if len(config.Entrypoint) == 1 {
+					// shell form ENTRYPOINT overrides shell form CMD so use it
+					// first.
+					s.Process.Args = nil
+					s.Process.CommandLine = config.Entrypoint[0]
+				} else if len(config.Cmd) == 1 {
+					// If no ENTRYPOINT then CMD should be used.
+					s.Process.Args = nil
+					s.Process.CommandLine = config.Cmd[0]
+				} else {
+					return errors.New("ImageConfig must contain either Entrypoint or Cmd when ArgsEscaped=true")
+				}
+				s.Process.Args = nil
+				if len(args) > 0 {
```

Comment:
I think we're missing the case where `len(Entrypoint) == 1`, `len(Cmd) == 1`, and `len(args) == 0`. In this case we should treat `Cmd` as `args` and escape + append it to `CommandLine`.

Reply:
I'm actually not sure if we should be doing this when `len(args) == 0` or only when actually `args == nil`. This seems unclear from the [image spec](https://github.com/opencontainers/image-spec/blob/main/config.md#properties). In CRI they have some [logic inherited from Docker](https://github.com/containerd/containerd/blob/main/pkg/cri/opts/spec.go#L62-L67) where an empty array can override the image's Entrypoint but not the Cmd...

Reply:
Unfortunately this is really unclear. Docker only has this test: https://github.com/moby/moby/blob/master/integration-cli/docker_cli_build_test.go#L5331 which actually shows it for escaping `CMD` (which the current pr supports). However, on Windows the current version of Docker actually would already not work to do what the code `inherited` from Docker supposedly solves. IE: 

```Dockerfile
FROM mcr.microsoft.com/windows/servercore:ltsc2019
ENTRYPOINT "C:\my path\test.exe"
CMD "-arg1 test"
```

```json
{
  "Config": {
    "Cmd": [
      "cmd /S /C \"-arg1 test\""
    ],
    "ArgsEscaped": true,
    "Entrypoint": [
      "cmd /S /C \"C:\\my path\\test.exe\""
    ]
  }
}
```

Because docker is `ArgsEscaping` with the `prependedshell` argument we can't even append `CMD` to `Entrypoint` by default. This would work for additionally passed in args like the PR here solves but we can't do:

```
CommandLine = Entrypoint[0] + " " + Cmd[0]
```

When `len(userAgs) == 0` and `len(Cmd) != 0`

How should this be handled?

Reply:
(Pardon the wall of text...)

After thinking about this some more I think I have an idea of what to do here. First some background/framing:

Let's start by acknowledging that Docker's use of ArgsEscaped is suboptimal. It allows for some more flexibility when specifying Windows command lines, but it's underspecified what the exact behavior should be. Also, as you show above, it doesn't really solve all the cases that might be desired.

So my proposal is that we don't focus as much on trying to make ArgsEscaped optimally solve the end-user scenario, but rather focus on duplicating Docker's support as much as possible. That will at least preserve the ability for images to run as expected if they were created with Docker. Longer-term we can focus on solving the overall scenario the "right" way with an addition to the OCI spec.

So let's look at how exactly Docker handles ArgsEscaped today:

When Docker goes to run an image, it has potentially 4 sets of arguments contributing to the runtime spec construction:
- Image entrypoint
- Image cmd
- Container entrypoint
- Container cmd

Docker first reconciles these by allowing the container entrypoint/cmd to override the image values ([link](https://github.com/moby/moby/blob/a7ea29a5a6dcf65a2f4fff0be9becc2f7c587b4e/daemon/commit.go#L68-L76)). Next it combines the entrypoint/cmd together into a single slice, and separates out the first element (`Path`) from the rest (`Args`) ([link 1](https://github.com/moby/moby/blob/5e62ca1a05f2aab49066d91c9070cb7dc7ad2461/daemon/container.go#L192-L197), [link 2](https://github.com/moby/moby/blob/5e62ca1a05f2aab49066d91c9070cb7dc7ad2461/daemon/container.go#L154-L155)).

Finally, if ArgsEscaped is set, Docker takes the `Path` value as is, then escapes each arg in `Args` and joins them all with `" "`.

In effect what this means is that only the first argument in the list will ever be treated as already escaped. Even if you wrote a Dockerfile that has shell form `ENTRYPOINT` and `CMD` directives, only the entrypoint will actually be used as-is.

_So all we would need to do then_ is determine if we override image args with container args, then shove entrypoint and cmd together, treat the first arg as already escaped if ArgsEscaped is set, and we're good to go.

_However_ there is one aspect of the Docker behavior that I don't think we should carry over: Because Docker overrides image entrypoint/cmd with container values before considering ArgsEscaped, the ArgsEscaped value can actually affect the interpretation of the overriding container entrypoint/cmd values. This means that if ArgsEscaped is set, and you completely override the image values by passing in your own, Docker will actually treat the first element of your args as if it is already escaped. This is a problem, as ArgsEscaped should really only affect the behavior of the image values, and not the overall evaluation of args for the container.

To work around this, we need to track if the first arg in our list actually comes from the image or not. We only treat it as pre-escaped if it came from the image, otherwise we handle it normally. The case where the first arg is from the image can be expressed as when the image entrypoint is non-empty (since image entrypoint is never overridden here), or when the image cmd was not overridden.

I think the logic here would look like this:


```golang
cmd := config.Cmd
cmdFromImage := true
if len(args) > 0 {
	cmd = args
	cmdFromImage = false
}

cmd = append(config.Entrypoint, cmd...)
if len(cmd) == 0 {
	return errors.New("No arguments specified")
}

if argsEscaped && (len(config.Entrypoint) > 0 || cmdFromImage) {
	s.Process.Args = nil
	s.Process.CommandLine = cmd[0] + " " + escapeArgs(cmd[1:])
} else {
	s.Process.Args = cmd
	s.Process.CommandLine = ""
}
```

Reply:
Done.